src/autoschedulers/mullapudi2016/AutoSchedule.cpp

```diff
@@ -1300,6 +1300,38 @@ class GPUTilingDedup {
         debug(2) << ")\n";
     }
 
+    /** Ensure the ordering of the inner and outer variables.
+     *
+     * Used as a bubble sorting algorithm of the "nested_parallelism" algorithm.
+     * The nested parallelism always ensures the inner and outer vars are
+     * adjacent to each order when this function is called.
+     */
+    void ensure_ordering(const VarOrRVar &inner, const VarOrRVar &outer) {
+        debug(2) << f.name() << ".ensure_ordering(" << inner.name() << ", " << outer.name() << ")\n";
+
+        const auto findItem = [&](const VarOrRVar &v) {
+            return std::find_if(ordering.begin(), ordering.end(), [v_name = v.name()](const VarOrRVar &item) {
+                return item.name() == v_name;
+            });
+        };
+
+        auto inner_iter = findItem(inner);
+        auto outer_iter = findItem(outer);
+
+        internal_assert(inner_iter != ordering.end());
+        internal_assert(outer_iter != ordering.end());
+
+        // The nested parallelism implements a bubble sorting algorithm, which
+        // ensures the inner and outer variables are adjacent to each other.
+        // Assert the requirement here.
+        internal_assert(std::abs(std::distance(inner_iter, outer_iter)) == 1);
+
+        if (inner_iter > outer_iter) {
+            std::iter_swap(inner_iter, outer_iter);
+        }
+        is_initial_order = false;
+    }
+
```

Comment:
> This call is highly suspicious. The other calls set an order for every dimension. But this only reorders two vars within the existing order in the else branch below. When this is called here, it discards other dimensions that might be present.

>  I don't want to prescribe an implementation strategy. At a minimum, I want to be confident the behaviors match and don't regress, even if the actual behavior is less useful or efficient than it could be.

@alexreinking I implemented `GPUTilingDedup::ensure_ordering` to reflect the separate behavior: reorder the adjacent two vars within the existing order. Appreciate if you can take a look at it.

Reply:
Re: function renaming.

@mcourteaux I called this function `require_ordering` simply for me to grep the debug log. I can rename it to `reorder` if you like. Please :+1: if so.

Reply:
I don't understand. The code changes seem to indicate you named it `reorder`; and not `require_ordering`. `reorder` is fine for me.